src/box/memtx_engine.cc

```diff
@@ -1093,10 +1110,17 @@ checkpoint_f(va_list ap)
 		if (checkpoint_write_invalid_system_row(snap) != 0)
 			goto fail;
 	});
-	if (checkpoint_write_raft(snap, &ckpt->raft) != 0)
-		goto fail;
-	if (checkpoint_write_synchro(snap, &ckpt->synchro_state) != 0)
-		goto fail;
+	/*
+	 * There may be no user data (e.g. when only RAFT_PROMOTE is written),
+	 * write limbo and raft states right after system spaces, at the end
+	 * of the checkpoint.
+	 */
+	if (!is_synchro_written) {
+		if (checkpoint_write_raft(snap, &ckpt->raft) != 0)
+			goto fail;
+		if (checkpoint_write_synchro(snap, &ckpt->synchro_state) != 0)
+			goto fail;
```

Comment:
Let's put these two function calls in a new helper function (`checkpoint_write_system_data` or something like that).

Reply:
Done